multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,159 @@
+import org.apache.spark.sql.{DataFrame, SaveMode, SparkSession}
+import org.apache.spark.sql.functions.{callUDF, col, lit, map}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import java.util.Arrays;
+
+object MultipassAddressing {
+  def main(args: Array[String]): Unit = {
+    val inputAddressPath = args(0)
+    val resourcesLocation = args(1)
+    val dataLocation = args(2)
+    val downloadLocation = args(3)
+    val outputDirectory = args(4)
+
+    val sparkConf = new SparkConf()
+    sparkConf.setIfMissing("spark.master", "local[*]")
+    val session = SparkSession.builder()
+      .config(sparkConf)
+      .getOrCreate()
+
+    // Load the addresses from the csv
+    val addressInputDF = AddressInput.open(session, inputAddressPath)
+
+    // set preference to return all fields
+    val prefBuilder = new PreferencesBuilder().withReturnAllInfo(true).build()
+
+    val downloadManager = new DownloadManagerBuilder(downloadLocation)
+      .addDownloader(new S3Downloader(session.sparkContext.hadoopConfiguration))
+      .addDownloader(new HDFSDownloader(session.sparkContext.hadoopConfiguration))
+      .addDownloader(new LocalFilePassthroughDownloader())
+      .build()
+
+    // build a AddressingUDF, with custom executor
+    val addressingUdf: UserDefinedFunction = new AddressingBuilder()
+      .withResourcesLocation(resourcesLocation)
+      .withDataLocations(dataLocation)
+      .withDownloadManager(downloadManager)
+      .udfBuilder()
+      .withPreferences(prefBuilder)
+      .withOutputFields(
+        "location.feature.geometry.coordinates.x as X",
+        "location.feature.geometry.coordinates.y as Y",
+        "address.formattedStreetAddress as formattedStreetAddress",
+        "address.formattedLocationAddress as formattedLocationAddress",
+        "customFields['PRECISION_CODE'] as PRECISIONCODE")
+      .withErrorField("error")
+      .forCustomExecutor(new CustomExecutor())
+
+    // call UDF for each row in the address DataFrame, this will result in the dataframe containing a new column for each
+    // of the specified output fields as well as an error column
+    addressInputDF
+      // Adds a new column, represented as a collection comprised of the outputFields and the error field
+      .withColumn("addressing_geocode_result", addressingUdf(map(
+        lit("street"), col("streetName"),
+        lit("city"), col("city"),
+        lit("admin1"), col("state"),
+        lit("postalCode"), col("zip")))
+      )
+      // Persist the addressing result to avoid recalculation when we expand the result
+      .persist()
+      // Expand the result collection such that each output field is a separate column, including the error field.
+      .select("*", "addressing_geocode_result.*").drop("addressing_geocode_result")
+      // Write the dataframe to the specified output folder as parquet file
+      .write.mode(SaveMode.Overwrite).option("header","true").csv(outputDirectory)
```

Comment:
We should do what the comment says to do and save this file in parquet format.

Reply:
i'd rather change the comment :)
Input is in CSV and since its a sample - i'd rather write a csv for easy inspection

Reply:
Ok, I have changed the comment to csv

Reply:
We should be storing the output in Parquet format.  This was decided when we did the first sample.

Reply:
@chris-allison-precisely Is there any specific reason that's why we decided to store it in parquet format!? Otherwise, csv is good as it is human-readable.

Reply:
It was originally Ray's argument to store it in Parquet.  I believe the thought was along the lines that was likely what the users should be using as a data format, not csv.  However, if he no longer feels that way, then I'm fine with leaving it as CSV.

Reply:
me too